README.md
# Perseus Size Optimization Plugin

> Note: users of v0.1.0-v0.1.3 should upgrade to v0.1.4 if they're using Rust 2021, as an upstream bug occurs if you attempt to compile your app in release mode. See [here](#compiling-in-release-mode-never-finishes) for further details.

This is a very simple plugin for [Perseus](https://arctic-hen7.github.io/perseus) that applies size optimizations automatically, which
decrease the size of your final Wasm bundle significantly, meaning faster loads for users because a smaller amount of data needs to be
transferred to their browsers. Because Perseus renders a page almost immediately through static generation, and then the Wasm bundle is
needed to make that page interactive, applying this plugin will decrease your app's time-to-interactive and its total blocking time
(a Lighthouse metric that normal Perseus apps don't do so well in on mobile).

If you're new to Perseus, check it out [on it's website](https://arctic-hen7.github.io/perseus) and [on GitHub](https://github.com/arctic-hen7/perseus)! Basically though, it's a really fast and fully-featured web framework for Rust!

## Usage

In your `src/lib.rs`, call the following function on `PerseusApp`:

``` rust
PerseusApp::new()
    [...]
    .plugins(Plugins::new()
        .plugin(
            perseus_size_opt,
            SizeOpts::default()
        ))
```

<details>
<summary>I'm using `define_app!`</summary>

If you're still using `define_app!` from v0.3.3, you should upgrade to using `PerseusApp` soon, but you can still use this plugin by adding the following to the bottom of the `define_app!` call:

```rust
plugins: Plugins::new().plugin(perseus_size_opt, SizeOpts::default())
```

</details>

If you have any other plugins defined, add the `.plugin()` call where appropriate. You'll also need to add the following imports:

```rust
use perseus_size_opt::{perseus_size_opt, SizeOpts};
```

Once that's done, run `perseus tinker` to apply the optimizations (this needs a separate command because they involve modifying the `.perseus/` directory), and then you can work with your app as normal!

If you ever want to uninstall the plugin, just remove the relevant `.plugin()` call and re-run `perseus tinker`, and it'll be completely removed.

## Optimizations

This plugin currently performs the following optimizations:

- `wee_alloc` -- an alternative allocator designed for Wasm that reduces binary size at the expense of slightly slower allocations
- `lto` -- reduces binary size when set to `true` by enabling link-time optimizations
- `opt-level` -- optimizes aggressively for binary size when set to `z`
- `codegen-units` -- makes faster and smaller code when set to lower values (but makes compile times slower in release mode)

Note that all optimizations will only apply to release builds. except for the use of `wee_alloc`, which will also affect development builds.

## Options

There are a few defaults available for setting size optimization levels, or you can build your own custom settings easily by constructing
`SizeOpts` manually.

- `::default()` -- enables all optimizations
- `::default_no_lto()` -- enables all optimizations except `lto = true`, because that can break compilation of execution on some hosting providers, like Netlify.
- `::only_wee_alloc()` -- only uses `wee_alloc`, applying no other optimizations
- `::no_wee_alloc()` -- applies all optimizations other than `wee_alloc`

## Known Bugs

### Compiling in release mode never finishes

[This](https://github.com/arctic-hen7/perseus/issues/83) is due to an upstream issue in Rust 2021 that leads to size optimizations of the `fluent-bundle` package causing an overload of LLVM (see [this issue](https://github.com/rust-lang/rust/issues/91011)). As of v0.1.4, this plugin accounts for this and does not attempt to optimize `fluent-bundle` with the `default()` options. However, this will increase bundle size, so it's recommended that, until this upstream issue is fixed, users of this plugin remain on Rust 2018 for now and use the `default_2018()` function instead. This will optimize `fluent-bundle` appropriately while ensuring that this bug does not occur. If moving to Rust 2018 is infeasible, you'll have to put up with slightly larger bundles for now until the upstream issue is fixed (this seems to depend on the LLVM team now).

## Stability

This plugin is considered quite stable due to how basic its optimizations are (the whole thing is one file), and so its stability is mostly dependent on that of Perseus. If you're happy to use Perseus, you shouldn't need to worry about using this plugin as well (in fact, it's recommended that all Perseus apps use this plugin).

## License

See [`LICENSE`](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.9](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.8...v0.1.9) (2022-04-20)


### Features

* made compatible with perseus `^0.3.4` ([986ab44](https://github.com/arctic-hen7/perseus-size-opt/commit/986ab44c73b43b3c6b6fca2315bf730ca25370fd))

### [0.1.8](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.7...v0.1.8) (2022-04-20)


### Features

* upgraded for perseus v0.3.4 ([bcfbc68](https://github.com/arctic-hen7/perseus-size-opt/commit/bcfbc68ace29b7fd15f1ed7ef70dc7930cb65b45))

### [0.1.7](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.6...v0.1.7) (2021-12-21)

### [0.1.6](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.5...v0.1.6) (2021-12-20)

### [0.1.5](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.4...v0.1.5) (2021-12-13)


### Features

* updated to support latest perseus/sycamore ([081544a](https://github.com/arctic-hen7/perseus-size-opt/commit/081544a7a9d7ab4040415847036a9dd3bf3a29e0))

### [0.1.4](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.3...v0.1.4) (2021-12-02)


### Bug Fixes

* added patch for `fluent-bundle` ([130222b](https://github.com/arctic-hen7/perseus-size-opt/commit/130222b7596082a79d0973f17b3990e66ddfd4b6))

### [0.1.3](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.2...v0.1.3) (2021-11-29)

### [0.1.2](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.1...v0.1.2) (2021-11-07)

### [0.1.1](https://github.com/arctic-hen7/perseus-size-opt/compare/v0.1.0...v0.1.1) (2021-10-30)


### Documentation Changes

* 📝 added readme warning about Perseus [#66](https://github.com/arctic-hen7/perseus-size-opt/issues/66) ([8fafa43](https://github.com/arctic-hen7/perseus-size-opt/commit/8fafa43f6931e161f606e01bfcf1be5f703e3fc8))


### Code Refactorings

* ✏️ changed title of example from `Tiny` to `Simple` ([#2](https://github.com/arctic-hen7/perseus-size-opt/issues/2)) ([a6b192b](https://github.com/arctic-hen7/perseus-size-opt/commit/a6b192b0a170908049516186f319e611a03948e6))
* 👽️ updated for latest perseus version ([989d235](https://github.com/arctic-hen7/perseus-size-opt/commit/989d235b21eea45bc8a0273f3a9a248933db7330)), closes [#1](https://github.com/arctic-hen7/perseus-size-opt/issues/1)

## 0.1.0 (2021-10-19)


### Features

* ✨ added plugin functionality and docs ([37c6a07](https://github.com/arctic-hen7/perseus-size-opt/commit/37c6a07242c3787641d36d09bb749d0b8ab5d9f8))
* 🎉 added basic oss boilerplate ([f3c8fa5](https://github.com/arctic-hen7/perseus-size-opt/commit/f3c8fa54255bcaa6b5a55024f45c668623bc116b))


### Bug Fixes

* 📝 fixed link in readme ([afe4481](https://github.com/arctic-hen7/perseus-size-opt/commit/afe448143e9fd6765e3a900fa69a18c45dec733f))


src/lib.rs
/*!
This is a very simple plugin for [Perseus](https://arctic-hen7.github.io/perseus) that applies size optimizations automatically, which
decrease the size of your final Wasm bundle significantly, meaning faster loads for users because a smaller amount of data needs to be
transferred to their browsers. Because Perseus renders a page almost immediately through static generation, and then the Wasm bundle is
needed to make that page interactive, applying this plugin will decrease your app's time-to-interactive and its total blocking time
(a Lighthouse metric that normal Perseus apps don't do so well in on mobile).

If you're new to Perseus, check it out [on it's website](https://arctic-hen7.github.io/perseus) and [on GitHub](https://github.com/arctic-hen7/perseus)! Basically though, it's a really fast and fully-featured web framework for Rust!

## Usage

In your `src/lib.rs`, add the following to the bottom of the `define_app!` macro:

```rust
plugins: Plugins::new().plugin(perseus_size_opt(), SizeOpts::default())
```

If you have any other plugins defined, add the `.plugin()` call where appropriate. You'll also need to add the following imports:

```rust
use perseus_size_opt::{perseus_size_opt, SizeOpts};
```

Once that's done, run `perseus tinker` to apply the optimizations (this needs a separate command because they involve modifying the `.perseus/` directory), and then you can work with your app as normal!

If you ever want to uninstall the plugin, just remove the relevant `.plugin()` call and re-run `perseus tinker`, and it'll be completely removed.

## Optimizations

This plugin currently performs the following optimizations:

- `wee_alloc` -- an alternative allocator designed for Wasm that reduces binary size at the expense of slightly slower allocations
- `lto` -- reduces binary size when set to `true` by enabling link-time optimizations
- `opt-level` -- optimizes aggressively for binary size when set to `z`
- `codegen-units` -- makes faster and smaller code when set to lower values (but makes compile times slower in release mode)

Note that all optimizations will only apply to release builds. except for the use of `wee_alloc`, which will also affect development builds.

## Options

There are a few defaults available for setting size optimization levels, or you can build your own custom settings easily by constructing
`SizeOpts` manually.

- `::default()` -- enables all optimizations
- `::default_no_lto()` -- enables all optimizations except `lto = true`, because that can break compilation of execution on some hosting providers, like Netlify.
- `::only_wee_alloc()` -- only uses `wee_alloc`, applying no other optimizations
- `::no_wee_alloc()` -- applies all optimizations other than `wee_alloc`

## Stability

This plugin is considered quite stable due to how basic its optimizations are (the whole thing is one file), and so its stability is mostly dependent on that of Perseus. If you're happy to use Perseus, you shouldn't need to worry about using this plugin as well (in fact, it's recommended that all Perseus apps use this plugin).
*/

use cargo_toml::{Dependency, Manifest, Profile, Value};
use perseus::plugins::{empty_control_actions_registrar, Plugin, PluginAction, PluginEnv};
use perseus::Html;
use std::collections::BTreeMap;
use std::fs;
use thiserror::Error;
use toml::value::Map;

const PLUGIN_NAME: &str = "perseus-size-opt";
const WEE_ALLOC_DEF: &str = "#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;";
// This will need updating over time
const WEE_ALLOC_VERSION: &str = "0.4";

/// Options for size optimizations. Note that these settings will only affect release builds (e.g. when you run `perseus deploy`), not
/// development builds (except using `wee_alloc`, whcih will affect everything).
pub struct SizeOpts {
    /// Whether or not to use `wee_alloc`, which reduces binary size significantly. This defaults to `true`. Note that this also makes
    /// allocations slightly slower, so you'll need to decide what tradeoff between size and speed you want your app to have.
    pub wee_alloc: bool,
    /// Whether or not to use link-time optimizations, which defaults to `true`. If your app isn't recognized by providers like Netlify,
    /// disable this.
    pub lto: bool,
    /// The optimization level to use, which defaults to `z` (aggressively optimize for speed). Cargo typically sets this to `3` for
    /// release builds.
    pub opt_level: String,
    /// The value for the `codegen-units` property, which is set by default to 1 by this plugin. Higher values here will mean faster
    /// compile times but slower code. The Rust default is 16 (256 with incremental builds).
    pub codegen_units: u16,
    /// Whether or not to enable the patch for `fluent-bundle` that fixes Perseus #83 (compiling taking forever with size optimizations). If you're running
    /// Rust 2021, you should enable this until [this upstream issue](https://github.com/rust-lang/rust/issues/91011) is fixed.
    pub enable_fluent_bundle_patch: bool,
}
impl Default for SizeOpts {
    fn default() -> Self {
        Self {
            wee_alloc: true,
            lto: true,
            opt_level: "z".to_string(),
            codegen_units: 1,
            enable_fluent_bundle_patch: true,
        }
    }
}
// We add a few more sensible named defaults
impl SizeOpts {
    /// The usual default, but without the `fluent-bundle` patch. Use this for greater size reductions if you're not using Rust 2021.
    pub fn default_2018() -> Self {
        Self {
            wee_alloc: true,
            lto: true,
            opt_level: "z".to_string(),
            codegen_units: 1,
            enable_fluent_bundle_patch: false,
        }
    }
    /// The usual default, but without `lto` enabled, which is known to cause problems on some hosting services like Netlify. If your
    /// app runs out of memory during compilation, or won't be served properly, try this.
    pub fn default_no_lto() -> Self {
        Self {
            wee_alloc: true,
            lto: false,
            opt_level: "z".to_string(),
            codegen_units: 1,
            enable_fluent_bundle_patch: true,
        }
    }
    /// Only enables the alternative allocator `wee_alloc`, with no further additional optimizations made.
    pub fn only_wee_alloc() -> Self {
        Self {
            wee_alloc: true,
            lto: false,
            opt_level: "3".to_string(),
            codegen_units: 16,
            enable_fluent_bundle_patch: true,
        }
    }
    /// Enables all optimizations other than changing the default allocator.
    pub fn no_wee_alloc() -> Self {
        Self {
            wee_alloc: false,
            lto: true,
            opt_level: "z".to_string(),
            codegen_units: 1,
            enable_fluent_bundle_patch: true,
        }
    }
}

/// The errors that this plugin can return.
#[derive(Error, Debug)]
pub enum Error {
    #[error("couldn't get and parse `.perseus/Cargo.toml`, try running `perseus tinker` again (without the `--no-clean` option)")]
    GetManifestFailed {
        #[source]
        source: cargo_toml::Error,
    },
    #[error("couldn't update `.perseus/Cargo.toml`, try running `perseus tinker` again (without the `--no-clean` option)")]
    WriteManifestFailed {
        #[source]
        source: std::io::Error,
    },
    #[error("couldn't read `.perseus/src/lib.rs`, try running `perseus tinker` again (without the `--no-clean` option)")]
    ReadLibFailed {
        #[source]
        source: std::io::Error,
    },
    #[error("couldn't update `.perseus/src/lib.rs`, try running `perseus tinker` again (without the `--no-clean` option)")]
    WriteLibFailed {
        #[source]
        source: std::io::Error,
    },
}

/// The actual mechanics of this plugin, which apply size optimizations to `.perseus/Cargo.toml` and `.perseus/src/lib.rs`.
fn apply_size_opts(opts: &SizeOpts) -> Result<(), Error> {
    // Get the internal `Cargo.toml` file in `.perseus/` (the current directory)
    let mut manifest = Manifest::from_path("Cargo.toml")
        .map_err(|err| Error::GetManifestFailed { source: err })?;
    // Apply size optimizations to the release profile
    let mut release_profile = manifest.profile.release.unwrap_or(
        // Because `Default` is not implemented for this `struct`...
        Profile {
            opt_level: None,
            lto: None,
            debug: None,
            rpath: None,
            debug_assertions: None,
            codegen_units: None,
            panic: None,
            incremental: None,
            overflow_checks: None,
            package: std::collections::BTreeMap::default(),
            build_override: None,
        },
    );
    release_profile.opt_level = Some(opts.opt_level.clone().into());
    release_profile.lto = Some(opts.lto.into());
    release_profile.codegen_units = Some(opts.codegen_units); // If the `fluent-bundle` patch is enabled, apply it
                                                              // TODO Remove this patch entirely once the upstream issue in LLVM is fixed and the error no longer occurs
    if opts.enable_fluent_bundle_patch {
        let mut fluent_bundle_conf = Map::new();
        fluent_bundle_conf.insert("opt-level".to_string(), Value::Integer(2));
        let mut patch = BTreeMap::new();
        patch.insert(
            "fluent-bundle".to_string(),
            Value::Table(fluent_bundle_conf),
        );
        release_profile.package = patch;
    }
    manifest.profile.release = Some(release_profile);
    // Add `wee_alloc` as a dependency if we're using that optimization
    if opts.wee_alloc {
        // We override any existing versions of `wee_alloc`, the user can disable that optimization if they're doing more advanced stuff
        manifest.dependencies.insert(
            "wee_alloc".to_string(),
            Dependency::Simple(WEE_ALLOC_VERSION.to_string()),
        );
    }
    // For some reason, our modifications to the manifest wipe out the `cdylib` definition (TODO investigate this further), so we need to add it back
    let mut manifest_lib = manifest.lib.unwrap_or_default();
    manifest_lib.crate_type = Some(vec!["cdylib".to_string(), "rlib".to_string()]);
    manifest.lib = Some(manifest_lib);

    // Write the new manifest
    // This will result in a ton of previously implied stuff being written, so the manifest will get a lot longer (fine because it's internal)
    let manifest_str = toml::to_string(&manifest).unwrap();
    fs::write("Cargo.toml", manifest_str)
        .map_err(|err| Error::WriteManifestFailed { source: err })?;

    if opts.wee_alloc {
        // Again, this is inside `.perseus/`, we're modifying the engine, not the user's code
        let lib_contents =
            fs::read_to_string("src/lib.rs").map_err(|err| Error::ReadLibFailed { source: err })?;
        //
        // Prepend the new allocator definition to the file
        // We actually need to put it on the second line after a module-level attribute
        let lib_contents_with_wee_alloc = lib_contents.replace(
            "#![allow(clippy::unused_unit)]",
            &format!("#![allow(clippy::unused_unit)]\n{}\n", WEE_ALLOC_DEF),
        );

        fs::write("src/lib.rs", lib_contents_with_wee_alloc)
            .map_err(|err| Error::WriteLibFailed { source: err })?;
    }

    Ok(())
}

/// Gets the plugin itself to be handed to Perseus' `define_app!` macro. Note that this plugin's optimizations will only take effect
/// in release mode (e.g. when you run `perseus deploy`).
pub fn perseus_size_opt<G: Html>() -> Plugin<G, SizeOpts> {
    Plugin::new(
        PLUGIN_NAME,
        |mut actions| {
            actions
                .tinker
                .register_plugin(PLUGIN_NAME, |_, plugin_data| {
                    if let Some(plugin_data) = plugin_data.downcast_ref::<SizeOpts>() {
                        let res = apply_size_opts(plugin_data);
                        if let Err(err) = res {
                            panic!("error in `perseus-size-opt`: {}", err);
                        }
                    } else {
                        unreachable!();
                    }
                });
            actions
        },
        empty_control_actions_registrar,
        // This plugin only needs to run at tinker-time, otherwise it increases binary sizes
        PluginEnv::Server,
    )
}
